### PR TITLE
Add support for incremental repair

### LINKED ISSUE #31: Re-add support for incremental repairs
As part of implementing concurrent virtual node repairs the ability to run incremental repairs was temporarily removed. Incremental repairs should be added again before 1.0.0 release.

### PR DESCRIPTION
**Pull request by masokol**
Closes #31

### CODE REVIEW

File: standalone-integration/src/test/java/com/ericsson/bss/cassandra/ecchronos/standalone/ITIncrementalOnDemandRepairJob.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    @Parameterized.BeforeParam
+    public static void init(Boolean remoteRoutingOption) throws IOException
+    {
+        myRemoteRouting = remoteRoutingOption;
+        initialize();
+
+        mockTableRepairMetrics = mock(TableRepairMetrics.class);
+
+        Node localNode = getNativeConnectionProvider().getLocalNode();
+        skipIfCassandraVersionLessThan4(localNode);
```
Comment (masokol) [inline standalone-integration/src/test/java/com/ericsson/bss/cassandra/ecchronos/standalone/ITIncrementalOnDemandRepairJob.java:110]: This is not needed once #494 is done.

File: standalone-integration/src/test/java/com/ericsson/bss/cassandra/ecchronos/standalone/ITIncrementalSchedules.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    private Set<TableReference> myRepairs = new HashSet<>();
+
+    @Parameterized.BeforeParam
+    public static void init(Boolean remoteRoutingOption) throws IOException
+    {
+        myRemoteRouting = remoteRoutingOption;
+        initialize();
+
+        Node localNode = getNativeConnectionProvider().getLocalNode();
+        skipIfCassandraVersionLessThan4(localNode);
```
Comment (masokol) [inline standalone-integration/src/test/java/com/ericsson/bss/cassandra/ecchronos/standalone/ITIncrementalSchedules.java:111]: This is not needed once #494 is done.

**Review by tommystendahl — APPROVED** [review]

File: application/src/test/resources/multiple_schedules_regex.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#
+# Copyright 2020 Telefonaktiebolaget LM Ericsson
```
Comment (itskarlsson) [inline application/src/test/resources/multiple_schedules_regex.yml:2]: 2023

File: application/src/test/resources/multiple_schedules.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#
+# Copyright 2020 Telefonaktiebolaget LM Ericsson
```
Comment (itskarlsson) [inline application/src/test/resources/multiple_schedules.yml:2]: 2023

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/IncrementalOnDemandRepairJob.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+    @Override
+    public void finishJob()
+    {
+        UUID id = getId();
+        getOnFinishedHook().accept(id);
+        if (myTasks.isEmpty())
+        {
+            getOngoingJob().finishJob();
+            LOG.info("Completed On Demand Repair: {}", id);
```
Comment (itskarlsson) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/IncrementalOnDemandRepairJob.java:118]: Incremental On Demand Repair

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/IncrementalRepairJob.java [hunk 6]
```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright 2018 Telefonaktiebolaget LM Ericsson
```
Comment (itskarlsson) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/IncrementalRepairJob.java:2]: 2023

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairGroup.java [hunk 7]
```diff
@@ -107,14 +109,14 @@ public boolean execute()
...
-
             try
             {
                 repairTask.execute();
             }
             catch (ScheduledJobException e)
             {
-                LOG.warn("Encountered issue when running repair task {}", repairTask, e);
+                LOG.warn("Encountered issue when running repair task {}, {}", repairTask, e.getMessage());
+                LOG.debug("", e);
```
Comment (itskarlsson) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairGroup.java:119]: I don't believe this is supposed to be here

File: docs/ECCTOOL_EXAMPLES.md [hunk 8]
```diff
@@ -8,12 +8,13 @@ In this example, we will use `ecctool repairs` to check the status of manual rep
...
-| f4fb2b38-b9d0-4390-97ca-eeb284391f80 | ee32d9c7-1a4e-40c2-9b28-1000544011ae | test     | table1 | IN_QUEUE  | 0.00        | -                   |
-| f4fb2b38-b9d0-4390-97ca-eeb284391f80 | ba7665b2-5a7b-42b5-9f38-037f2da1e80a | test     | table1 | COMPLETED | 100.00      | 2022-09-22 13:40:07 |
----------------------------------------------------------------------------------------------------------------------------------------------------
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------
+| Id                                   | Host Id                              | Keyspace | Table  | Status    | Repaired(%) | Completed at        | Repair type |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------
+| f4fb2b38-b9d0-4390-97ca-eeb284391f80 | ee32d9c7-1a4e-40c2-9b28-1000544011ae | test     | table1 | IN_QUEUE  | 0.00        | -                   | VNODE       |
```
Comment (itskarlsson) [inline docs/ECCTOOL_EXAMPLES.md:14]: What do you think of having two tables under the same command?

Two tables that are printed out by the command. One for incremental one for vnode. If either incremental or vnode doesn't exist we could just not print it. This way we don't get a long list of repair types. What do you think?

**Review by itskarlsson — CHANGES_REQUESTED** [review]

File: docs/ECCTOOL_EXAMPLES.md (hunk 8 above)
Comment (masokol) [reply docs/ECCTOOL_EXAMPLES.md:14]: Hmm, i need to think about this.

Comment (masokol) [reply docs/ECCTOOL_EXAMPLES.md:14]: I think having 2 tables will be confusing for the users if --limit flag is used. Maybe we should add --incremental --vnode flags to "schedules" and "repairs" to filter on repair type, wdyt?

Comment (itskarlsson) [reply docs/ECCTOOL_EXAMPLES.md:14]: I understand what you mean but unfortunately that wouldn't really fix the readability of it all for new users. I agree that limit does make splitting a bit of an issue and I can accept having flags instead to filter on repair type.

**Review by itskarlsson — APPROVED** [review]

**Pull request merged**